packages/jest-mock/src/__tests__/index.test.ts

```diff
@@ -1089,6 +1098,83 @@ describe('moduleMocker', () => {
     });
   });
 
+  describe('mockWhen', () => {
+    it('mocks specific args', () => {
+      const fn = moduleMocker.fn();
+      fn.mockReturnValue('generic output');
+      fn.mockWhen('special input').mockReturnValue('special output');
+
+      expect(fn('arbitrary')).toBe('generic output');
+      expect(fn('special input')).toBe('special output');
+    });
+
+    it('works with several whens', () => {
+      const fn = moduleMocker.fn();
+      fn.mockImplementation(num => `${num}teen`);
+      fn.mockWhen('one').mockReturnValue('eleven');
+      fn.mockWhen('two').mockReturnValue('twelve');
+      fn.mockWhen('three').mockReturnValue('thirteen');
+
+      expect(fn('one')).toBe('eleven');
+      expect(fn('two')).toBe('twelve');
+      expect(fn('three')).toBe('thirteen');
+      expect(fn('four')).toBe('fourteen');
+    });
+
+    it('works without a default implementation', () => {
+      const fn = moduleMocker.fn();
+      fn.mockWhen('special input').mockReturnValue('special output');
+
+      expect(fn('arbitrary')).toBeUndefined();
+      expect(fn('special input')).toBe('special output');
+    });
+
+    it('supports matchers', () => {
+      const fn = moduleMocker.fn();
+      fn.mockReturnValue('hello');
+      fn.mockWhen(expect.any(String)).mockImplementation(v => `hello, ${v}`);
+      fn.mockWhen('hello', expect.any(String)).mockImplementation(
+        (_, y) => `morning, ${y}`,
+      );
+
+      expect(fn(3)).toBe('hello');
+      expect(fn('world')).toBe('hello, world');
+      expect(fn('hello', 'friend')).toBe('morning, friend');
```

Comment:
Maybe add a negative partial matching case.
```suggestion      
      expect(fn('hello', 'friend')).toBe('morning, friend');
      expect(fn('hello', 5)).toBe('hello');
```